src/scripts/syncBreaches.js

```diff
@@ -11,7 +11,81 @@
 
 import { req, formatDataClassesArray } from '../utils/hibp.js'
 import { getAllBreaches, upsertBreaches } from '../db/tables/breaches.js'
+import { mkdir, readdir } from "node:fs/promises";
+import { resolve as pathResolve } from "node:path";
+import { finished } from 'node:stream/promises';
+import fs from "node:fs";
+import { Readable } from 'node:stream';
+import aws from 'aws-sdk';
 
+// Get breaches logos and uploads to s3
+const accessKeyId = process.env.AWS_ACCESS_KEY_ID;
+const secretAccessKey = process.env.AWS_SECRET_ACCESS_KEY;
+const region = process.env.AWS_REGION;
+const Bucket = process.env.S3_BUCKET;
+
+const s3 = new aws.S3({
+  region,
+  credentials: {
+    accessKeyId,
+    secretAccessKey,
+  },
+});
+
+async function uploadToS3(fileName, fileStream) {
+  console.log('Attempt to upload to s3: ', fileName)
+  const uploadParams = {
+    Bucket,
+    Key: fileName,
+    Body: fileStream
+  }
+  try {
+    await s3.upload(uploadParams).promise()
+    console.log('Successfully uploaded data to ' + Bucket + '/' + fileName)
+  } catch (err) {
+    console.error(err, err.stack)
+  }
+}
+
+export async function getBreachIcons(breaches) {
+  let logoMap;
+  const breachDomains = breaches
+    .map((breach) => breach.Domain)
+    .filter((breachDomain) => breachDomain.length > 0);
+
+  // make logofolder if it doesn't exist
+  const logoFolder = "public/logo_cache/";
+  if (!fs.existsSync(logoFolder)) await mkdir(logoFolder);
+
+  // read existing logos
+  const existingLogos = await readdir(logoFolder);
+  console.log(`existing logos: ${existingLogos.length}`)
+
+  // TODO: Batch to limit memory use?
+  const logoMapElems = (await Promise.all(
+    breachDomains.map(async (breachDomain) => {
+      const logoFilename = breachDomain.toLowerCase() + ".ico";
+      const logoPath = pathResolve(logoFolder, logoFilename);
+      if (existingLogos.includes(logoFilename)) {
+        console.log('skipping ', logoFilename)
+        return;
+      }
+      console.log(`fetching: , ${logoFilename}`)
+      const res = await fetch(
+        `https://icons.duckduckgo.com/ip3/${breachDomain}.ico`);
+      await uploadToS3(logoFilename, Buffer.from(await res.arrayBuffer()))
```

Comment:
Does S3 care if the file is already there?
If I'm following this correctly the script will check if files exist LOCALLY and if not, it will fetch them from DDG and then upload to S3. So I guess I'm mildly curious on whether a file is missing locally but possibly already exists in S3, will S3 throw errors.

I guess the alternative is to not worry about local assets and query S3 bucket if a logo already exists and if so, skip and if not, upload. Although that comes at the expense of an extra 670ish GET requests to S3 to see if images exist (and 99.999999% they will).

Or conversely, if we're running the cron every... 6-24 hours, we could just filter the breaches by records that have a `ModifiedDate` that is around the same cron period. Possibly not perfect, but means 99.9999% of the time we don't need to fetch or check-for _any_ images in S3 if there haven't been any changes to the hibp breaches feed.

Reply:
Or here's a rough version where I scraped the S3 bucket XML's listing and scanned for missing logos: https://github.com/pdehaan/blurts-s3-logos
Hopefully there is a better method to check S3 using the S3 API versus me scraping XML, but hey... I only actually needed 2 network requests to see which breaches from HIBP were missing from the S3 bucket (one for HIBP /breaches and the other for the [S3 bucket](http://s3.amazonaws.com/firefoxmonitor-dev-monitor-cdn-dev-static-website/)).

Spoiler alert: No logos were missing in S3, so I had to fake an item into the `breachDomains` Set to see if my logo checking was working.

Reply:
I think it makes sense to overwrite existing logos, since companies might update their logos anyway, so we probably want the newest versions.

Reply:
S3 upload will always overwrite / replace. S3 has a concept of versioning that's built in so the old version is preserved by default.. though we can configure it (SRE) with different retention policies etc. Yea I think we are ok doing an overwrite here